Reactions and their observables simplified

chrisfroe opened the pull request:
- reaction names are unique
- reaction class is template free
- kernel specific state-models know reaction counts -> `kernel.initialize()` properly sets them to contain a zero for every registered reaction (otherwise the reaction-counts will have `size()==0`, when observables are recorded first)

chrisfroe committed:
[context] reaction names are unique

chrisfroe committed:
simplified Reaction class

chrisfroe committed:
move type definition and init method of reaction counts into separate
header file

chrisfroe committed:
Base StateModel knows reactionCounts, such that kernel.initialize() can properly initialize them.
Remove reaction index from ReactionInfo and ReactionRecord, global ids are used for identification, user only sees names if possible.
Flattened out order1/order2 structure in ReactionCounts observable.

clonker commented on include/readdy/model/reactions/Utils.h:
```diff
@@ -0,0 +1,58 @@
+
+using reaction_counts_map = std::unordered_map<reactions::Reaction::reaction_id, std::size_t>;
+
+inline void zeroReactionCounts(reaction_counts_map &reactionCounts,
```
I think this should rather be a method of the state model impl and take the reaction registry instead of a map. Also I propse calling the method `resetReactionCounts` or something alike

clonker commented on include/readdy/model/reactions/Utils.h:
```diff
@@ -0,0 +1,58 @@
+
+inline void zeroReactionCounts(reaction_counts_map &reactionCounts,
+                               const std::unordered_map<Reaction::reaction_id, Reaction *> &allReactions) {
+    reactionCounts.clear();
```
is the clear necessary?

clonker commented on readdy/main/model/reactions/ReactionRegistry.cpp:
```diff
@@ -413,21 +405,89 @@ ReactionRegistry::addDecay(const std::string &name, const std::string &type, sca
+}
+
+const ReactionRegistry::reactions_raw_ptr_map ReactionRegistry::allReactions() const {
+    reactions_raw_ptr_map all;
```
I would call this `idMap` (or something alike). Also i would not call this method within the simulation loop as it makes copies out of the `orderNFlat` methods which in itself make copies already. Instead, one could use `registry.order1()` and `registry.order2()` directly.

clonker commented on include/readdy/model/reactions/Utils.h:
```diff
@@ -0,0 +1,58 @@
+                               const std::unordered_map<Reaction::reaction_id, Reaction *> &allReactions) {
+    reactionCounts.clear();
+    for (const auto &reaction : allReactions) {
+        const auto &r = reaction.second;
```
I would prefer an implementation that uses `registry.order1()` and `registry.order2()` instead.

clonker commented on kernels/cpu/src/actions/reactions/CPUGillespie.cpp:
```diff
@@ -52,7 +52,7 @@ void CPUGillespie::perform(const util::PerformanceNode &node) {
 
     if(ctx.recordReactionCounts()) {
-        readdy::model::observables::ReactionCounts::initializeCounts(stateModel.reactionCounts(), ctx);
+        readdy::model::reactions::utils::zeroReactionCounts(stateModel.reactionCounts(), ctx.reactions().allReactions());
```
this probably should be a method inside the `CPUStateModel`.

clonker commented on kernels/cpu/src/actions/reactions/CPUUncontrolledApproximation.cpp:
```diff
@@ -127,7 +126,7 @@ void CPUUncontrolledApproximation::perform(const util::PerformanceNode &node) {
     }
     if (ctx.recordReactionCounts()) {
-        readdy::model::observables::ReactionCounts::initializeCounts(stateModel.reactionCounts(), ctx);
+        readdy::model::reactions::utils::zeroReactionCounts(stateModel.reactionCounts(), ctx.reactions().allReactions());
```
this probably should be a method inside the `CPUStateModel`.

clonker commented on readdy/main/model/Kernel.cpp:
```diff
@@ -168,6 +168,9 @@ bool Kernel::supportsTopologies() const {
 void Kernel::initialize() {
     context().configure(true);
+    if (context().recordReactionCounts()) {
+        reactions::utils::zeroReactionCounts(stateModel().reactionCounts(), context().reactions().allReactions());
```
move method into state model

clonker commented on readdy/main/model/reactions/Reactions.cpp:
```diff
@@ -77,6 +77,79 @@ std::ostream& operator<<(std::ostream& os, const ReactionRecord& record) {
+
 }
 }
 }
```
add newline

clonker commented on wrappers/python/src/cxx/common/Utils.cpp:
```diff
@@ -287,13 +287,15 @@ std::vector<std::vector<rpy::ReadableReactionRecord>> read_reactions_obs(const s
-    std::vector<readdy::model::ioutils::ReactionInfo> reactionInfoOrder1;
-    std::vector<readdy::model::ioutils::ReactionInfo> reactionInfoOrder2;
+    // get reaction info from config
+    std::unordered_map<readdy::model::reactions::Reaction::reaction_id, readdy::model::ioutils::ReactionInfo> reactionsMap;
```
reformat please :)

chrisfroe committed:
resetReactionCounts() is a method of the kernel-specific StateModel classes, removed reactionCounts() from base class.
Type of reactionCounts is defined along side the ReactionRecords, removed the extra header.
Added missing order1() accessor in ReactionRegistry.
